chainer/optimizers/adam.py

```diff
@@ -68,6 +69,65 @@ def _inplace_axpby(x, a, b, y):
             x[:] = a * x + b * y
 
 
+class AdamKernel(object):
+    def __init__(self, hp):
+        self.hp = copy.copy(hp)
+        self.hp_dict = hp.get_dict()
+
+        @cuda.fuse(kernel_name='adam')
+        def kernel(grad, alpha_t, one_minus_beta1, one_minus_beta2, eps,
+                   eta, one_minus_weight_decay_rate, param, m, v, vhat):
+            xp = cuda.get_array_module(grad)
+            dtype = _get_intermediate_dtype(param.dtype.type)
+            grad = grad.astype(dtype, copy=False)
+
+            # m += (1 - beta1) * (grad - m)
+            _inplace_axpby(m, 1.0, one_minus_beta1, grad - m)
+            # v += (1 - beta2) * (grad * grad - v)
+            _inplace_axpby(v, 1.0, one_minus_beta2, grad * grad - v)
+
+            if vhat is None:
+                vhat = v
+            else:
+                xp.maximum(vhat, v, out=vhat)
+            vhat = vhat.astype(dtype, copy=False)
+
+            # param -= eta * (alpha_t * m / (sqrt(vhat) + eps)
+            #                 - weight_decay_rate * param)
+            _inplace_axpby(
+                param, one_minus_weight_decay_rate, -eta,
+                alpha_t * m / (xp.sqrt(vhat) + hp.eps))
+
+        def kernel_nonconst_hp(hp, grad, alpha_t, param, m, v, vhat=None):
```

Comment:
I feel the code structure is too difficult to understand.
Is there a chance for refactoring to make it easier?
Comments are needed, at the very least.